### ISSUE TITLE
Telegram (windows 10 v2.8.5 beta) Not remembering window position on close. 

### ISSUE THREAD

**Issue by bzarky** [telegramdesktop/tdesktop#16569]
been having an issue lately, past few months. Telegram does not remember the position of the window anymore. It consistently opens on the wrong monitor every time. I can drag it back to the spot its supposed to be, but when i close telegram and open it again, it opens back up on a different monitor. This is not exclusive to version 2.8.5 beta. Its a minor annoyance but it is an annoyance nonetheless, and is an unexpected change in behavior.

**Comment by ilya-fedin** [comment]
You mean close to the tray or full close?

**Comment by bzarky** [comment]
Fully quit telegram

**Comment by ilya-fedin** [comment]
> Fully quit telegram

enable debug logs via `debugmode` cheat code (go to settings screen and type the word blindly, make sure you have english layout activated), you'll get a dialog to activate debug logs. Then try to quit & launch Telegram, you should have log files in `%APPDATA%\Telegram Desktop\DebugLogs`, open the one with right timestamp and starting with `log`, the lines with `Window Pos:` are related, copy them and paste here

**Comment by bzarky** [comment]
OK I moved the window position to where it should be, quit telegram, and reopened it. Here are all the [Window Pos] Lines from my debuglog. My monitors are one 3840x2160 primary monitor, and one 1920x1080 secondary monitor. Telegram should be opening in the position on the big primary monitor, but keeps opening on the small secondary monitor. 
```
[20:50:30.613 00-0000255] Window Pos: Initializing first -962, 487, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[20:50:30.613 00-0000256] Window Pos: Screen found, screen geometry: 3840, 589, 1920, 1080 (main_window.cpp : 25)
[20:50:30.613 00-0000257] Window Pos: Resulting geometry is 3840, 614, 958, 1055 (main_window.cpp : 25)
[20:50:30.613 00-0000258] Window Pos: Setting first 3840, 614, 958, 1055 (main_window.cpp : 25)
[20:50:31.847 00-0000855] Window Pos: Saving non-maximized position: 3840, 614, 958, 1055 (main_window.cpp : 25)
[20:50:31.847 00-0000856] Window Pos: Screen found, geometry: 3840, 589, 1920, 1080 (main_window.cpp : 25)
[20:50:31.847 00-0000857] Window Pos: Writing: 0, 25, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[20:50:32.847 00-0000879] Window Pos: Writing to storage 0, 25, 958, 1055 (scale 100%, maximized [FALSE]) (core_settings.cpp : 25)
```

**Comment by ilya-fedin** [comment]
Here's a problem. `Window Pos: Writing:` lines should be first since it's writing resolution data to disk. `Window Pos: Initializing` lines should be chosen the ones that after the `Writing` lines, from the next launch.

**Comment by bzarky** [comment]
Here I moved the window again, quit, and reopened it again. and left it in the spot it opened at. 


`[20:54:29.444 00-0001041] Window Pos: Saving non-maximized position: 2882, 1077, 958, 1055 (main_window.cpp : 25)
[20:54:29.444 00-0001042] Window Pos: Screen found, geometry: 3840, 589, 1920, 1080 (main_window.cpp : 25)
[20:54:29.444 00-0001043] Window Pos: Writing: -958, 488, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[20:54:30.445 00-0001044] Window Pos: Writing to storage -958, 488, 958, 1055 (scale 100%, maximized [FALSE]) (core_settings.cpp : 25)
[20:54:31.389 00-0001045] Window Pos: Saving non-maximized position: 2879, 1077, 958, 1055 (main_window.cpp : 25)
[20:54:31.389 00-0001046] Window Pos: Screen found, geometry: 3840, 589, 1920, 1080 (main_window.cpp : 25)
[20:54:31.389 00-0001047] Window Pos: Writing: -961, 488, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[20:54:32.390 00-0001048] Window Pos: Writing to storage -961, 488, 958, 1055 (scale 100%, maximized [FALSE]) (core_settings.cpp : 25)
[20:54:33.981 00-0001049] Window Pos: Saving non-maximized position: 2881, 1077, 958, 1055 (main_window.cpp : 25)
[20:54:33.981 00-0001050] Window Pos: Screen found, geometry: 3840, 589, 1920, 1080 (main_window.cpp : 25)
[20:54:33.981 00-0001051] Window Pos: Writing: -959, 488, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[20:54:34.982 00-0001484] Window Pos: Writing to storage -959, 488, 958, 1055 (scale 100%, maximized [FALSE]) (core_settings.cpp : 25)
[20:58:39.765 00-0000255] Window Pos: Initializing first -959, 488, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[20:58:39.765 00-0000256] Window Pos: Screen found, screen geometry: 3840, 589, 1920, 1080 (main_window.cpp : 25)
[20:58:39.765 00-0000257] Window Pos: Resulting geometry is 3840, 614, 958, 1055 (main_window.cpp : 25)
[20:58:39.765 00-0000258] Window Pos: Setting first 3840, 614, 958, 1055 (main_window.cpp : 25)
[20:58:40.982 00-0000856] Window Pos: Saving non-maximized position: 3840, 614, 958, 1055 (main_window.cpp : 25)
[20:58:40.982 00-0000857] Window Pos: Screen found, geometry: 3840, 589, 1920, 1080 (main_window.cpp : 25)
[20:58:40.982 00-0000858] Window Pos: Writing: 0, 25, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[20:58:41.983 00-0000880] Window Pos: Writing to storage 0, 25, 958, 1055 (scale 100%, maximized [FALSE]) (core_settings.cpp : 25)
[20:50:30.613 00-0000255] Window Pos: Initializing first -962, 487, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[20:50:30.613 00-0000256] Window Pos: Screen found, screen geometry: 3840, 589, 1920, 1080 (main_window.cpp : 25)
[20:50:30.613 00-0000257] Window Pos: Resulting geometry is 3840, 614, 958, 1055 (main_window.cpp : 25)
[20:50:30.613 00-0000258] Window Pos: Setting first 3840, 614, 958, 1055 (main_window.cpp : 25)
[20:50:31.847 00-0000855] Window Pos: Saving non-maximized position: 3840, 614, 958, 1055 (main_window.cpp : 25)
[20:50:31.847 00-0000856] Window Pos: Screen found, geometry: 3840, 589, 1920, 1080 (main_window.cpp : 25)
[20:50:31.847 00-0000857] Window Pos: Writing: 0, 25, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[20:50:32.847 00-0000879] Window Pos: Writing to storage 0, 25, 958, 1055 (scale 100%, maximized [FALSE]) (core_settings.cpp : 25)`

**Comment by ilya-fedin** [comment]
```
[20:54:33.981 00-0001049] Window Pos: Saving non-maximized position: 2881, 1077, 958, 1055 (main_window.cpp : 25)
[20:54:33.981 00-0001050] Window Pos: Screen found, geometry: 3840, 589, 1920, 1080 (main_window.cpp : 25)
[20:54:33.981 00-0001051] Window Pos: Writing: -959, 488, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[20:54:34.982 00-0001484] Window Pos: Writing to storage -959, 488, 958, 1055 (scale 100%, maximized [FALSE]) (core_settings.cpp : 25)
```

looks like something is wrong with the logic that tries to find the right screen, I think it's better to wait @john-preston, he is the one who wrote it

**Comment by bzarky** [comment]
OK thanks for the help. It used to perform correctly, probably 3 or 4 months ago. I don't remember precisely.

**Comment by ilya-fedin** [comment]
That's weird, that logic seem to be the same since first Telegram versions

**Comment by bzarky** [comment]
`[20:54:33.981 00-0001050] Window Pos: Screen found, geometry: 3840, 589, 1920, 1080 (main_window.cpp : 25)`
Am I reading this correctly? does it think my primary monitor is only 589 pixels tall?  Or is this something else.

**Comment by ilya-fedin** [comment]
It's `x, y, width, height` coordinates of the screen

**Comment by bzarky** [comment]
So from what i can gather, its saving the correct window position here: 
`[20:54:33.981 00-0001049] Window Pos: Saving non-maximized position: 2881, 1077, 958, 1055 (main_window.cpp : 25)`

but then doing something completely different after.

**Comment by ilya-fedin** [comment]
`Saving non-maximized position` says the geometry it got from Qt
`Screen found, geometry` says geometry of the screen it found with some custom logic. As you can see, it finds a FHD screen here, the wrong one.
`Writing` and `Writing to storage` say the digits that are written to disk.

**Comment by bzarky** [comment]
OK some potentially new information that may be a bug in the program. 

If i drag the window far enough away from the right edge of my main monitor, it remembers the window location upon quit and reopen. 
If i have it too close to the right edge of my main monitor, it will put the window on my secondary monitor upon quit and reopen. 

These are the log lines from that: 

```
[22:52:07.017 00-0000255] Window Pos: Initializing first 286, 333, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[22:52:07.017 00-0000256] Window Pos: Screen found, screen geometry: 0, 0, 3840, 2160 (main_window.cpp : 25)
[22:52:07.017 00-0000257] Window Pos: Resulting geometry is 286, 333, 958, 1055 (main_window.cpp : 25)
[22:52:07.017 00-0000258] Window Pos: Setting first 286, 333, 958, 1055 (main_window.cpp : 25)
[22:52:08.193 00-0000861] Window Pos: Saving non-maximized position: 286, 333, 958, 1055 (main_window.cpp : 25)
[22:52:08.193 00-0000862] Window Pos: Screen found, geometry: 0, 0, 3840, 2160 (main_window.cpp : 25)
[22:52:12.739 00-0000884] Window Pos: Saving non-maximized position: 292, 336, 958, 1055 (main_window.cpp : 25)
[22:52:12.739 00-0000885] Window Pos: Screen found, geometry: 0, 0, 3840, 2160 (main_window.cpp : 25)
[22:52:12.739 00-0000886] Window Pos: Writing: 292, 336, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[22:52:13.740 00-0000887] Window Pos: Writing to storage 292, 336, 958, 1055 (scale 100%, maximized [FALSE]) (core_settings.cpp : 25)
[22:52:15.746 00-0000888] Window Pos: Saving non-maximized position: 997, 811, 958, 1055 (main_window.cpp : 25)
[22:52:15.746 00-0000889] Window Pos: Screen found, geometry: 0, 0, 3840, 2160 (main_window.cpp : 25)
[22:52:15.746 00-0000890] Window Pos: Writing: 997, 811, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[22:52:16.746 00-0000891] Window Pos: Writing to storage 997, 811, 958, 1055 (scale 100%, maximized [FALSE]) (core_settings.cpp : 25)
[22:52:24.908 00-0000255] Window Pos: Initializing first 997, 811, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[22:52:24.908 00-0000256] Window Pos: Screen found, screen geometry: 0, 0, 3840, 2160 (main_window.cpp : 25)
[22:52:24.908 00-0000257] Window Pos: Resulting geometry is 997, 811, 958, 1055 (main_window.cpp : 25)
[22:52:24.908 00-0000258] Window Pos: Setting first 997, 811, 958, 1055 (main_window.cpp : 25)
[22:52:26.083 00-0000442] Window Pos: Saving non-maximized position: 997, 811, 958, 1055 (main_window.cpp : 25)
[22:52:26.083 00-0000443] Window Pos: Screen found, geometry: 0, 0, 3840, 2160 (main_window.cpp : 25)
[22:52:36.923 00-0000891] Window Pos: Saving non-maximized position: 2881, 1076, 958, 1055 (main_window.cpp : 25)
[22:52:36.923 00-0000892] Window Pos: Screen found, geometry: 3840, 545, 1920, 1080 (main_window.cpp : 25)
[22:52:36.923 00-0000893] Window Pos: Writing: -959, 531, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[22:52:37.923 00-0000899] Window Pos: Writing to storage -959, 531, 958, 1055 (scale 100%, maximized [FALSE]) (core_settings.cpp : 25)
[22:52:42.681 00-0000255] Window Pos: Initializing first -959, 531, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[22:52:42.681 00-0000256] Window Pos: Screen found, screen geometry: 3840, 545, 1920, 1080 (main_window.cpp : 25)
[22:52:42.681 00-0000257] Window Pos: Resulting geometry is 3840, 570, 958, 1055 (main_window.cpp : 25)
[22:52:42.681 00-0000258] Window Pos: Setting first 3840, 570, 958, 1055 (main_window.cpp : 25)
[22:52:43.892 00-0000838] Window Pos: Saving non-maximized position: 3840, 570, 958, 1055 (main_window.cpp : 25)
[22:52:43.892 00-0000839] Window Pos: Screen found, geometry: 3840, 545, 1920, 1080 (main_window.cpp : 25)
[22:52:43.892 00-0000840] Window Pos: Writing: 0, 25, 958, 1055 (scale 100%, maximized [FALSE]) (main_window.cpp : 25)
[22:52:44.892 00-0000880] Window Pos: Writing to storage 0, 25, 958, 1055 (scale 100%, maximized [FALSE]) (core_settings.cpp : 25)
```

**Comment by ilya-fedin** [comment]
> If i drag the window far enough away from the right edge of my main monitor, it remembers the window location upon quit and reopen.

From what I see, the logic does some manipulation with center coordinates:
https://github.com/telegramdesktop/tdesktop/blob/c33be27b3c7a782509302120e80ad27240455694/Telegram/SourceFiles/window/main_window.cpp#L705-L716

**Comment by bzarky** [comment]
I was able to trick it into performing correctly by moving the position of my secondary monitor in windows display config, closing and reopening telegram with the window in the correct location, and it managed to stick this time. After, I was able to move my secondary monitor back to its previous location and, from what i can tell, telegram is now keeping the window in the correct location without spawning it onto my secondary monitor. 

This is a bit of a workaround but i guess its OK for now. I wonder what specifically what specifically was causing an issue, if it was how QT managed the screen geometry or how windows was reporting it.
